web/src/lib/auth/signinHelper.ts

```diff
@@ -52,9 +52,9 @@ export const onSelfRegister = ({
   }
   setRegistrationStatus("IN_PROGRESS");
 
-  const route = router.asPath?.includes(ROUTES.accountSetup)
-    ? ""
-    : getCurrentBusiness(userData).preferences.returnToLink || router.asPath || "";
+  const { returnToLink } = getCurrentBusiness(userData).preferences;
```

Comment:
These updated lines are the key to getting the returnToLink to actually work for a user who goes through self registration. It was completely resetting the returnToLink whenever a user went through the account setup page (which I think always happens, before self registration).